### ISSUE TITLE
more date functions

### ISSUE THREAD

**Issue by alfu32** [lodash/lodash#2714]
# Need
having functions hat would allow to specify and work with time intervals.

# Synopsis : 

each time point designated by a Date, has ( logically ) a resolution (or time span in absolute units ) - 

  `enum TimeSpan{year,quarter,month,week,day,workday,hour,minute,second or milisecond}`

an interval should have the following structure 
  ```
Interval{
     public Timespan unit;
     public Date begin;
     public Date end;
  }
```
# Use Cases : 
those functions should wrap cluttered code (see footnotes) into a semantically cleaner synthax

## 1 `_.date.forEach`
_subdivision iterator 

starting with only a time point (year, quarter, month, day, hour, minute, second)
to be able to iterate through its subdivisions, the iterator should not generate internally any arrays to support iteration ( if the time unit is big and division is small, it will swalow up a lot of memory)

```
_.date.now.forEach("day in year",function(Interval subdivision,Number index){
   // your cod here
})
_.date.forEach("day in month",function(start,end,index){
   // your cod here
})
```
more generally 
```
_.date.now.forEach(
  "TimeSpan <subdivision> in Timespan <interval>",
  function(Interval subdivision,Number index){
     // your cod here
  }
)
```
alternate synthax :

`_.date.<unit>.forEach(<subdivision>,function(interval,index){})`
examples
`_.date.year.forEach(month,function(interval,index){})`
`_.date.month.forEach(day,function(interval,index){})`

the function should throw an error if 

1. subdivision does not fit in interval
2. subdivision or interval are undefined TimeSpan names

## 2 `_.date.interval`
_interval from date
have a function that returns the interval from a Date (viewed as a timepoint)

`  _.date.new.interval(TimeSpan timeSpan):Interval`
examples :
`_.date.interval(new Date(2016,10,09,12,00,53),"day")`
should return 
`{unit:"day",begin:new Date(2016,10,09,0,0,0),end:new Date(2016,10,09,23,59,59)}`

`_.date.interval(new Date(2016,10,09,12,00,53),"day")`
should return 
`{unit:"day",begin:new Date(2016,10,09,0,0,0),end:new Date(2016,10,09,23,59,59)}`

## 3 `_.date.subdivide`
_generate collection of TimeSpans
begining with a timepoint, generate a collection or an array of subdivisions :

`_.date.subdivide(TimeSpan unit, Date : timepoint, TimeSpan subdivision):TimeSpan[]`

which in plain english should read for example : "subdivide into quarters year of 2016-09-10"
or "subdivide into days month of 2012-10-10"
more generally
"subdivide into <subdivision> <unit> of <timepoint>"

examples :

`_.date.subdivide(month,new Date(2016,10,09,12,32,53), day)`

### a `_.date.increment`

`_.date.inc(Date d,value,,unit=day):Date;`
examples 
`_.date.inc(new Date(2016,10,09,12,32,53),1,day) `should equal `new Date(2016,09,10,12,32,53)`
`_.date.inc(new Date(2016,10,31,12,32,53),1,day)` should equal `new Date(2016,11,01,12,32,53)`

### b `_.date.decrement`

_.date.dec(Date d,value,,unit=day):Date;

examples 
`_.date.dec(new Date(2016,09,09,12,32,53),1,day) `should equal `new Date(2016,09,08,12,32,53)`
`_.date.dec(new Date(2017,01,01,12,32,53),1,day)` should equal `new Date(2016,12,31,12,32,53)`

### c `_.date.normalize`

`_.date.norm(Date d,unit):Date;`

examples 
`_.date.norm(new Date(2016,10,09,12,32,53),day) `should equal `new Date(2016,09,08,12,32,53)`

### d `_.date.toInterval`
_( or  `_.date.interval` )

`_.date.toInterval(Date d,unit):Interval;`

examples 
`_.date.norm(new Date(2016,10,09,12,32,53),day) `should equal `new Date(2016,09,08,12,32,53)`

# Footnotes
the implementation of inc and dec is fairily simple :

```
//(you can replace day with any of the intervals mentioned in Synopsys)
/// input parameters
var t=new Date();
var increment=5;

/// increment day 
t.setDate(t.getDate()+increment)
/// it will correctly change year, and month if it is the case
/// it will take into account changing from summertime to wintertime
```

```
/// increment day 
/// input parameters
var fn=function(interval,index){}
/// inernal function stubs 
function normalize(date,interval){
   
}
function clone(date){
   return new Date(date);
}
function getInterval(date,intervalType){

}
/// code for foreach function
for(var index=0,d=startDate;d.getTime()<endDate.getTime();d.setDay(d.getDay()+1)){
  var interval=getInterval(clone(d),day);
   //your code here
  fn(clone(d),index);
  index++
}
```
(forEach could be based on inc,dec,normalize)

**Comment by jdalton** [comment]
Hi @alfu32!

I'm deferring additional date functions to the awesome [moment.js](https://github.com/moment/moment/)!

**Comment by StreetStrider** [comment]
@alfu32, BTW, moment has a [plugin](http://momentjs.com/docs/#/plugins/range/) to work with ranges: albegra, and even iterating. The iterating is not useful enough (as for me).
